### PR TITLE
Sync committee duties improvements&fixes

### PR DESCRIPTION
**Pull request by olegshmuelov**

### CODE REVIEW

File: beacon/goclient/duties.go [hunk 1]
```diff
@@ -70,8 +70,9 @@ func (gc *goClient) fetchAttesterDuties(epoch phase0.Epoch, validatorIndices []p
 	return duties, nil
 }
 
-// fetchProposerDuties applies proposer duties
-func (gc *goClient) fetchProposerDuties(epoch phase0.Epoch, validatorIndices []phase0.ValidatorIndex) ([]*spectypes.Duty, error) {
+// ProposerDuties applies proposer duties
+func (gc *goClient) ProposerDuties(epoch phase0.Epoch, validatorIndices []phase0.ValidatorIndex) ([]*spectypes.Duty, error) {
+	gc.logger.Debug("fetching proposer duties", zap.Uint64("epoch", uint64(epoch)), zap.Int("validatorIndices", len(validatorIndices)))
```
Comment (nivBlox) [inline beacon/goclient/duties.go:75]: im not sure we need this log

File: beacon/goclient/duties.go [hunk 2]
```diff
@@ -43,14 +42,15 @@ func (gc *goClient) GetDuties(epoch phase0.Epoch, validatorIndices []phase0.Vali
 	return duties, nil
 }
 
-// fetchAttesterDuties applies attester + aggregator duties
-func (gc *goClient) fetchAttesterDuties(epoch phase0.Epoch, validatorIndices []phase0.ValidatorIndex) ([]*spectypes.Duty, error) {
+// AttesterDuties applies attester + aggregator duties
+func (gc *goClient) AttesterDuties(epoch phase0.Epoch, validatorIndices []phase0.ValidatorIndex) ([]*spectypes.Duty, error) {
+	gc.logger.Debug("fetching attester duties", zap.Uint64("epoch", uint64(epoch)), zap.Int("validatorIndices", len(validatorIndices)))
```
Comment (nivBlox) [inline beacon/goclient/duties.go:47]: im not sure we need this log

File: operator/duties/controller.go [hunk 3]
```diff
@@ -143,17 +184,108 @@ func createDutyExecuteMsg(duty *spectypes.Duty, pubKey *bls.PublicKey) (*spectyp
...
+	if data.Slot != dc.ethNetwork.EstimatedCurrentSlot() {
+		return
+	}
+
+	// check for reorg
+	epoch := dc.ethNetwork.EstimatedEpochAtSlot(data.Slot)
+	if dc.lastBlockEpoch != 0 && epoch <= dc.lastBlockEpoch {
+		if !bytes.Equal(dc.currentDutyDependentRoot[:], zeroRoot[:]) &&
+			!bytes.Equal(dc.currentDutyDependentRoot[:], data.CurrentDutyDependentRoot[:]) {
+			dc.logger.Debug("Current duty dependent root has changed",
```
Comment (nivBlox) [inline operator/duties/controller.go:205]: do we need the root logged?

File: operator/duties/controller.go [hunk 4]
```diff
@@ -143,17 +184,108 @@ func createDutyExecuteMsg(duty *spectypes.Duty, pubKey *bls.PublicKey) (*spectyp
...
+}
+
 // listenToTicker loop over the given slot channel
 func (dc *dutyController) listenToTicker(slots <-chan phase0.Slot) {
 	for currentSlot := range slots {
-		// execute duties
+
+		if dc.ethNetwork.IsFirstSlotOfEpoch(currentSlot) {
+			currentEpoch := dc.ethNetwork.EstimatedEpochAtSlot(currentSlot)
+			indices := dc.validatorController.GetValidatorsIndices()
```
Comment (nivBlox) [inline operator/duties/controller.go:222]: the get indices might block a bit the execute duties, WDYT?

File: protocol/v2/blockchain/beacon/network.go [hunk 5]
```diff
@@ -67,3 +67,21 @@ func (n Network) IsFirstSlotOfEpoch(slot phase0.Slot) bool {
 func (n Network) GetEpochFirstSlot(epoch phase0.Epoch) phase0.Slot {
 	return phase0.Slot(epoch * 32)
 }
+
+// EstimatedSyncCommitteePeriodAtEpoch estimates the current sync committee period at the given Epoch
+func (n Network) EstimatedSyncCommitteePeriodAtEpoch(epoch phase0.Epoch) uint64 {
+	return uint64(epoch) / 256 // EpochsPerSyncCommitteePeriod
```
Comment (nivBlox) [inline protocol/v2/blockchain/beacon/network.go:73]: lets use a const instead of hard coded

File: protocol/v2/blockchain/beacon/network.go [hunk 6]
```diff
@@ -67,3 +67,21 @@ func (n Network) IsFirstSlotOfEpoch(slot phase0.Slot) bool {
...
+}
+
+// FirstEpochOfSyncPeriod calculates the first epoch of the given sync period.
+func (n Network) FirstEpochOfSyncPeriod(period uint64) phase0.Epoch {
+	return phase0.Epoch(period * 256) // EpochsPerSyncCommitteePeriod
```
Comment (nivBlox) [inline protocol/v2/blockchain/beacon/network.go:78]: lets use a const instead of hard coded

File: operator/duties/controller.go (hunk 4 above)
Comment (olegshmuelov) [reply operator/duties/controller.go]: we are using map, should be ok

File: protocol/v2/blockchain/beacon/network.go (hunk 5 above)
Comment (olegshmuelov) [reply protocol/v2/blockchain/beacon/network.go:73]: noted!
we have a dedicated task for this

**Review by nivBlox — APPROVED** [review]

**Review by moshe-blox — CHANGES_REQUESTED** [review]

**Review by moshe-blox — APPROVED** [review]

**Pull request merged**